src/common/bin_serde.rs

```diff
@@ -0,0 +1,99 @@
+//! Binary serialization, and an implementation over Unix pipes.
+use std::{
+    io::{self, Read, Write},
+    marker::PhantomData,
+    os::unix::net::UnixStream,
+};
+
+/// Serialization/deserialization trait using a byte array as storage.
+pub trait BinSerDe<const N: usize> {
```

Comment:
I understand now how a generic doesn't have the right intent. I would add that the API is cumbersome as we have to carry around sizes as well. 

Unfortunately an associated constant doesn't work:
```rs
pub trait BinSerDe {
    const LEN: usize;
    fn serialize(&self) -> [u8; Self::LEN];
    fn deserialize(bytes: [u8; Self::LEN]) -> Self;
}
```
as it is blocked by https://github.com/rust-lang/rust/issues/60551.

Moving it into a function:
```rs
pub trait BinSerDe {
    fn len() -> usize;
    fn serialize(&self) -> [u8; Self::len()];
    fn deserialize(bytes: [u8; Self::len()]) -> Self;
}
```
does not work because there is no `const fn` in traits.

If we switch to `&[u8]`, inducing a little overhead:
```rs
pub trait BinSerDe {
    const LEN: usize;
    fn serialize(&self) -> &[u8];
    fn deserialize(bytes: &[u8]) -> Self;
}
```

... and instead create the fixed size buffers in the `impl`:
```rs
pub fn read(&mut self) -> io::Result<R> {
    let mut bytes = [0u8; R::LEN]; // ...
}
```

it doesn't work, and [would need nightly to compile, for `feature(generic_const_exprs)`](https://stackoverflow.com/questions/65461672/rust-constant-expression-depends-on-a-generic-parameter), which [isn't stable yet](https://github.com/rust-lang/rust/issues/76560).

What do you think? If there isn't a good API that lets us keep the fixed size buffers, then I feel like it's more trouble than it's worth and we could close.